feat($sce): handle URLs through the $sce service

petebacondarwin opened the pull request:
Replaces #14250

Thanks to @rjamet for the original work on this feature.

This is a large patch to handle URLs with the $sce service, similarly to HTML context.

However, to keep rough compatibility with existing apps, we need to allow URL-context
concatenation, since previously $sce contexts prevented sanitization. There's now a
set of special contexts (defined in $interpolate) that allow concatenation, in a roughly
intuitive way:

* Trusted types alone are trusted, e.g. `"{{safeType}}"` will not be sanitized.
* Any concatenation of values results in a non-trusted type that will be handled by
`getTrusted` once the concatenation is done, e.g. `"{{ 'javascript:foo' }}"` and
`"javascript:{{safeType}}"` will be sanitized.

This commit also introduces a new SCE context called `SRC`, which represents a URL
being used as a source for an image, video, audio, etc. The hierarchy is setup
so that the `URL` context is also a `SRC` context, in the same way that the `RESOURCE_URL`
context is also a `URL` (and now also a `SRC`) context.

Where we previously sanitized URL attributes inside the compiler, we now only apply
the `$sce.URL` or `$sce.SRC` context requirement.

* When calling `getTrustedSrc()` a value that is not already a trusted `SRC` will be
sanitized using the `imgSrcSanitizationWhitelist`
* When calling `getTrustedUrl()` a value that is not already a trusted `URL` will be
sanitized using the `aHrefSanitizationWhitelist`

This results in behaviour that closely matches the previous sanitization behaviour.

<!-- General PR submission guidelines https://github.com/angular/angular.js/CONTRIBUTING.md#submit-pr -->
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

feature

**What is the current behavior? (You can also link to an open issue here)**

URLs are sanitized directly in the compiler or interpolator.

**What is the new behavior (if this is a feature change)?**

Now that sanitization is done via the SCE service.

**Does this PR introduce a breaking change?**
Yes

If you use `attrs.$set` for URL attributes there will be no automated sanitization
of the URL value. This is now in line with other contexts. If you are programmatically
writing URL values to attributes from untrusted input then you must sanitize
it yourself (possibly by calling the private `$$sanitizeUri` service).

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our [guidelines](../DEVELOPERS.md#commits)
- [x] Fix/Feature: [Docs](../DEVELOPERS.md#documentation) have been added/updated
- [x] Fix/Feature: Tests have been added; existing tests pass

**Other information**:

petebacondarwin commented on src/ng/directive/attrs.js:
```diff
@@ -436,7 +436,7 @@ forEach(['src', 'srcset', 'href'], function(attrName) {
           // then calling element.setAttribute('src', 'foo') doesn't do anything, so we need
           // to set the property as well to achieve the desired effect.
-          // We use attr[attrName] value since $set can sanitize the url.
+          // We use attr[attrName] value since $set might have sanitized the url.
```
Perhaps this comment is no longer valid, given the breaking change notice...

petebacondarwin replied:
Actually if the attribute is `srcset` then there can still be sanitization

jbedard commented on src/ng/interpolate.js:
```diff
@@ -283,27 +289,55 @@ function $InterpolateProvider() {
         }
       }
 
+      if (concat.length === 1 && expressionPositions.length === 1) {
```
Can we remove the initial `singleExpression` assignment and change this to `singleExpression = (concat.length === 1 && expressionPositions.length === 1)`?

jbedard commented on src/ng/interpolate.js:
```diff
@@ -283,27 +289,55 @@ function $InterpolateProvider() {
+              // doesn't allow that.
+              $interpolateMinErr.throwNoconcat(text);
+            } else {
+              return concat.join('');
```
Drop this (and the next `else`) and just fallthru? Could then combine the condition above to `trustedContext && concat.length > 1`...

jbedard commented on src/ng/interpolate.js:
```diff
@@ -340,8 +374,13 @@ function $InterpolateProvider() {
-          return allOrNothing && !isDefined(value) ? value : stringify(value);
+          if (contextAllowsConcatenation && singleExpression) {
+            // No stringification in this case, to keep the trusted value until unwrapping.
+            return value;
```
This basically makes `parseStringifyInterceptor` a noop, can we avoid `parseStringifyInterceptor` (and any interceptor) completely in that case? Maybe only if that case is common enough that it's worth avoiding interceptors...

petebacondarwin replied:
Difficulty is that we assign the interpolator to the parseFn before we have calculated how many items are in `contat` and `expressionPositions` so we don't know whether `singleExpression` is true until we have finished.

petebacondarwin replied:
But I managed it...

jbedard replied:
If it's difficult or ugly then we can leave it out of this PR and try later since it's not really directly related to this PR

rjamet commented on src/ng/sanitizeUri.js:
```diff
@@ -19,7 +20,8 @@ function $$SanitizeUriProvider() {
    * regular expression. If a match is found, the original url is written into the dom. Otherwise,
-   * the absolute url is prefixed with `'unsafe:'` string and only then is it written into the DOM.
+   * the absolute url is prefixed with `'unsafe:'` string and only then is it written into the DOM,
+   * making it inactive.
```
Could you maybe document that those are used in the $sce.URL sanitization now, so it's reflected in the API docs ? (and same below)

rjamet commented on src/ng/sce.js:
```diff
@@ -22,8 +22,11 @@ var SCE_CONTEXTS = {
 
-  // An URL used in a context where it does not refer to a resource that loads code. Currently
-  // unused in AngularJS.
+  // An URL used in a context where it refers to the source of media such as an image, audio, video, etc.
```
If you push it a little, iframes and scripts are media. I'd say "to the source of media which are not expected to run scripts such as ..." instead.

rjamet commented on src/ng/sce.js:
```diff
@@ -386,15 +390,27 @@ function $SceDelegateProvider() {
+     * the context and sanitizer availablility.
+     *
+     * The contexts that can be sanitized are $sce.MEDIA_URL, $sce.URL and $sce.HTML. The first two are available
+     * by default, and the second one relies on the $sanitize service (which may be loaded through
```
s/second/third/

rjamet commented on src/ng/sce.js:
```diff
@@ -386,15 +390,27 @@ function $SceDelegateProvider() {
-     * is available or possible.)
+     * This function will throw if the safe type isn't appropriate for this context, or if the
+     * value given cannot be accepted in the context (which might be caused by sanitization not
+     * being available, or the value being unsafe).
```
"the value not being recognized as safe" instead?

rjamet commented on src/ng/sce.js:
```diff
@@ -412,12 +428,19 @@ function $SceDelegateProvider() {
+      if (type === SCE_CONTEXTS.MEDIA_URL) {
+        return $$sanitizeUri(maybeTrusted, true);
+      } else if (type === SCE_CONTEXTS.URL) {
+        return $$sanitizeUri(maybeTrusted);
```
explicit false as a second argument there ?

petebacondarwin replied:
oh, OK then :-) I was saving bytes.

rjamet commented on src/ng/sanitizeUri.js:
```diff
@@ -44,7 +46,8 @@ function $$SanitizeUriProvider() {
-   * the absolute url is prefixed with `'unsafe:'` string and only then is it written into the DOM.
+   * the absolute url is prefixed with `'unsafe:'` string and only then is it written into the DOM,
+   * making it inactive.
    *
```
Below in this file, I would replace the isImage argument with isMediaUrl to be clearer. That shouldn't break anything.

rjamet commented on src/ng/sanitizeUri.js:
```diff
@@ -6,6 +6,7 @@
  */
 function $$SanitizeUriProvider() {
+
   var aHrefSanitizationWhitelist = /^\s*(https?|s?ftp|mailto|tel|file):/,
```
There's a tiny quirk hiding there, which is probably without consequences: mailto and tel are in the URL whitelist, but not in the MEDIA_URL. This makes sense, but "safe under URL" is supposed to be a subset of "safe under MEDIA_URL" in my interpretation of the subclassing in the $sce. I don't think anyone cares either way.

petebacondarwin replied:
Yeah, I wondered about that. One is not really a "super-set" of the other but I needed both to be covered by `RESOURCE_URL`, so I had to force a hierarchy on them. Given that users can manipulate these whitelists in anyway they like, I don't feel too bad about the defaults not quite aligning.

rjamet commented on src/ngSanitize/sanitize.js:
```diff
@@ -154,7 +153,7 @@ function $SanitizeProvider() {
       var buf = [];
       htmlParser(html, htmlSanitizeWriter(buf, function(uri, isImage) {
-        return !/^unsafe:/.test($$sanitizeUri(uri, isImage));
+        return !/^unsafe:/.test($$sanitizeUri(uri), isImage);
```
This looks wrong to me?

petebacondarwin replied:
but no test is failing 😱

petebacondarwin replied:
This is because I broke the tests in https://github.com/angular/angular.js/pull/16378/files#diff-85da78d5073ece90913c9e733cfcdef4 !!!

rjamet commented on test/ng/compileSpec.js:
```diff
@@ -11028,91 +11083,110 @@ describe('$compile', function() {
-          $rootScope.testUrl = $sce.trustAsUrl('http://example.com/image2.mp4');
+        it('should accept trusted values', inject(function($rootScope, $compile, $sce) {
+          // As a MEDIA_URL URL
+          element = $compile('<' + tag + ' src="{{testUrl}}"></' + tag + '>')($rootScope);
```
I think the CI errors you get are due to some browser protections on assigning javascript: to images... Those are annoying, I remember hitting these on the IE family.

petebacondarwin replied:
Right, I wondered about that. I'll remove the `javascript:`s

rjamet commented:
LGTM overall, nice work!

jbedard commented on src/ng/interpolate.js:
```diff
@@ -283,29 +284,40 @@ function $InterpolateProvider() {
       }
 
+      singleExpression = concat.length === 1 && expressionPositions.length === 1;
+      parseFns = contextAllowsConcatenation && singleExpression ?
```
Would this be any shorter/nicer doing...
```
parseFns = expressions.map(function(exp) {
    return $parse(exp, contextAllowsConcatenation && singleExpression ? undefined : parseStringifyInterceptor);
});
```

petebacondarwin replied:
I thought about that. It does mean evaluating that `contextAllowsConcatenation && singleExpression` for each item in `concat`. Also I was being cautious in case the `$parse` was doing parameter counting rather than just a check for `undefined`. If you feel strongly about it we could change it.

jbedard replied:
I think it reads a little nicer but don't feel strongly about it.

Could do `var interceptor = ...` outside the loop. That might also be a little more readable which I think is most important here (I don't think performance is a major concern since this isn't run per digest or anything like that).

petebacondarwin replied:
I tried putting the interceptor outside the loop but it fails because the `interceptor` variable somehow gets shared between closures and breaks the tests. I followed your original suggestion.

petebacondarwin replied:
Done

Narretz commented on src/ng/interpolate.js:
```diff
@@ -340,7 +352,13 @@ function $InterpolateProvider() {
-          value = getValue(value);
+          // In concatenable contexts, getTrusted comes at the end, to avoid sanitizing individual
+          // parts of a full URL. We don't care about losing the trustedness here.
+          // If non-concatenable contexts where there is only one expression then this interceptor
```
If > In?

Narretz commented on src/ng/sce.js:
```diff
@@ -594,17 +618,22 @@ function $SceDelegateProvider() {
+ * <div class="alert alert-warning">
+ * Be aware that `a[href]` and `img[src]` used to automatically sanitize their URLs and not pass them
+ * through {@link ng.$sce#getTrusted $sce.getTrusted}. **As of 1.7.0, this is no longer the case.**
+ * Now `getTrusted` will sanitize values for the `$sce.MEDIA_URL` and `$sce.URL` contexts.
```
It's not clear which context applies to what. How about "Now `getTrusted` will sanitize `a[href]` for the `$sce.URL` context, and `img[src]` for the `$sce.MEDIA?URL` context.

petebacondarwin replied:
This is not strictly related to DOM nodes. What this is saying is that, previously the `a[href]` and `img[src]` did not use $sce at all but did their own sanitization and that $sce did not do anything with the `$sce.URL` (and now also `$sce.MEDIA_URL`) contexts. Now the `a[href]` and `img[src]` do nothing but require the `URL` or `MEDIA_URL` context. It is `$sce` that now does the sanitizing as appropriate.

Narretz replied:
Ok, but the paragraph makes it sound like (at least to me) that the second part still refers to a[href] and img[src], and as such it's confusing that the last sentence isn't clear about which context gets applied to them. So I think adding the explicit context matching is a good idea.

petebacondarwin replied:
doc comments updated

Narretz commented on test/ng/sceSpecs.js:
```diff
@@ -525,6 +552,44 @@ describe('SCE', function() {
+      expect($sce.getTrustedUrl($sce.trustAsResourceUrl('javascript:foo'))).toEqual('javascript:foo');
+    }));
+
+    it('should use the $$sanitizeUri', function() {
```
Isn't this technically an implementation detail?

petebacondarwin replied:
Well I see it as an interface (albeit internal) to the service. Testing it this way means that we are less tied to the implementation of the `$$sanitizeUri` service.

Narretz reviewed (changes requested):
As far as I can say, this looks reasonable, although I find it a bit weird that the strategy for sanitizing content that is programmtically written is to use the private $$sanitizeUri service. This feels wrong. Maybe we should make the service public?

The commit message would also be clearer if this section:

> Where we previously sanitized URL attributes when setting attribute value inside the
`$compile` service, we now only apply the `$sce.URL` or `$sce.SRC` context requirement,
and leave the `$interpolate` service to deal with sanitization.

would come immediately after

> This is a large patch to handle URLs with the $sce service, similarly to HTML context.

Btw, 

>  we now only apply the `$sce.URL` or `$sce.SRC` context requirement

I don't think $sce.SRC is anywhere in this patch, is this line right?

gkalpak commented on src/ng/compile.js:
```diff
@@ -1710,13 +1710,20 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
-            (nodeName === 'img' && key === 'src') ||
-            (nodeName === 'image' && key === 'xlinkHref')) {
-          // sanitize a[href] and img[src] values
-          this[key] = value = $$sanitizeUri(value, nodeName === 'img' || nodeName === 'image');
```
So, doesn't this leave people using `ng-href/src` less vulnerable (unless they use interpolation)?
(Not common to use `ng-href/src` without interpollation, but possible.)

petebacondarwin replied:
I am not sure what you mean here. Do you mean "more" vulnerable?
It is true that if you were programmatically setting the attribute via `$set` then it is no longer checked, but this is quite a common paradigm, where one has to be responsible for coding against the lower level API.

gkalpak replied:
Yeah, I meant "more" (everybody knows "less" is "more" 😛)
The thing is that some of the built-in directives (e.g. `ngHref``/`ngSrc`) use `$set` under the hood.

petebacondarwin replied:
I am pretty certain that `ngHref` and `ngSrc` are not affected. They only call `$set` with a value that is passed from `$observe`, which has already interpolated the value, and so has passed the value through SCE.

gkalpak commented on docs/content/error/$compile/srcset.ngdoc:
```diff
@@ -0,0 +1,12 @@
+passing a "trusted" value provided by calls to `$sce.trustAsMediaUrl(value)`.
+
+If you want to programmatically set explicitly trusted unsafe URLs, you should use `$sce.trustAsHtml`
+on the whole `img` tag and inject it into the DOM using the `ng-bind-html` directive.
```
:scream:

petebacondarwin commented on src/ng/directive/attrs.js:
```diff
@@ -436,7 +436,7 @@ forEach(['src', 'srcset', 'href'], function(attrName) {
           // then calling element.setAttribute('src', 'foo') doesn't do anything, so we need
           // to set the property as well to achieve the desired effect.
-          // We use attr[attrName] value since $set can sanitize the url.
+          // We use attr[attrName] value since $set might have sanitized the url.
```
We still sanitize `img[srcset]` when called through `$set`.

petebacondarwin commented on src/ng/sanitizeUri.js:
```diff
@@ -59,10 +68,10 @@ function $$SanitizeUriProvider() {
-      var normalizedVal;
-      normalizedVal = urlResolve(uri && uri.trim()).href;
+    return function sanitizeUri(uri, isMediaUrl) {
+      if (!uri) return uri;
```
Hmm, you're right that `urlResolve` doesn't (mostly) return the empty string when passed an empty string... I need to look again at why I did this. There was definitely a reason.

petebacondarwin replied:
I can't remember (or see) why I did this. So I am going to revert that change.

petebacondarwin commented on src/ng/sce.js:
```diff
@@ -594,17 +619,27 @@ function $SceDelegateProvider() {
- * | `$sce.JS`           | For JavaScript that is safe to execute in your application's context.  Currently, no bindings require this context.  Feel free to use it in your own directives. |
+ * | `$sce.HTML`         | For HTML that's safe to source into the application.  The {@link ng.directive:ngBindHtml ngBindHtml} directive uses this context for bindings. If an unsafe value is encountered and the {@link ngSanitize $sanitize} module is present this will sanitize the value instead of throwing an error. |
+ * | `$sce.CSS`          | For CSS that's safe to source into the application.  Currently unused.  Feel free to use it in your own directives. |
+ * | `$sce.MEDIA_URL`    | For URLs that are safe to display as media. Is automatically converted from string by sanitizing when needed. |
```
I'm going with `render` it is vague enough to cover audio IMO.

petebacondarwin commented on test/ng/directive/booleanAttrsSpec.js:
```diff
@@ -152,16 +152,28 @@ describe('ngSrc', function() {
           'when a trusted value is required.  See http://docs.angularjs.org/api/ng.$sce');
   }));
 
+  it('should interpolate a multi-part expression for img src attribute (URL context)', inject(function($compile, $rootScope) {
```
MEDIA_URL - thanks. I also tidied up the description of the previous test

petebacondarwin commented on test/ng/directive/booleanAttrsSpec.js:
```diff
@@ -195,23 +207,23 @@ describe('ngSrc', function() {
 
           $rootScope.$digest();
-          expect(element.prop('src')).toEqual('some/');
+          expect(element.prop('src')).toContain('some/');
```
Nada

petebacondarwin replied:
changed to `toEqual`.

petebacondarwin replied:
On my machine the value is: `http://192.168.1.161:9876/some/` - in other words the protocol, host and port, etc get prefixed in front of the path.

petebacondarwin replied:
Changed to `expect(element.prop('src')).toMatch('/some/$');`

petebacondarwin commented on test/ng/directive/booleanAttrsSpec.js:
```diff
@@ -195,23 +207,23 @@ describe('ngSrc', function() {
             $rootScope.id = $sce.trustAsResourceUrl('http://somewhere');
           });
-          expect(element.prop('src')).toEqual('http://somewhere');
+          expect(element.prop('src')).toEqual('http://somewhere/');
```
It is adding the `/` because that is how the empty "path" is determined by the url resolver.
I changed it to 

```
$rootScope.$apply(function() {
            $rootScope.id = $sce.trustAsResourceUrl('http://somewhere/abc');
          });
          expect(element.prop('src')).toEqual('http://somewhere/abc');
```

which shows that we don't arbitrarily add a slash to the end of the path

petebacondarwin commented on test/ng/directive/booleanAttrsSpec.js:
```diff
@@ -294,9 +306,8 @@ describe('ngHref', function() {
-      element = $compile('<a ng-href="http://www.google.com/{{\'a%link\'}}">')($rootScope);
-
       expect(function() {
+        element = $compile('<a ng-href="http://www.google.com/{{\'a%link\'}}">')($rootScope);
```
stylistic - I reverted that change, for clarity.

petebacondarwin commented on test/ng/interpolateSpec.js:
```diff
@@ -380,33 +385,57 @@ describe('$interpolate', function() {
+
+
+    it('should interpolate and sanitize a multi-part expression when isTrustedContext is URL', inject(function($sce, $interpolate) {
+      /* jshint scripturl:true */
```
:-) picked up from @rjamet's original PR that was written before we moved to eslint. Removed

petebacondarwin commented on test/ng/directive/booleanAttrsSpec.js:
```diff
@@ -152,16 +152,28 @@ describe('ngSrc', function() {
+      $rootScope.$apply(function() {
+        $rootScope.id = 1;
+      });
       dealoc(element);
```
probably not 🙀

petebacondarwin replied:
But if the test failed it would be called, so it might be a good thing to clean up in that case?

gkalpak commented:
I am a little concerned about https://github.com/angular/angular.js/pull/16378#discussion_r164105466.
Other than that LGTM (as long as Travis is happy :wink:)!

petebacondarwin committed:
feat($sce): handle URL sanitization through the `$sce` service

Thanks to @rjamet for the original work on this feature.

This is a large patch to handle URLs with the $sce service, similarly to HTML context.

Where we previously sanitized URL attributes when setting attribute value inside the
`$compile` service, we now only apply an `$sce` context requirement and leave the
`$interpolate` service to deal with sanitization.

This commit introduces a new `$sce` context called `MEDIA_URL`, which represents
a URL used as a source for a media element that is not expected to execute code, such as
image, video, audio, etc.
The context hierarchy is setup so that a value trusted as `URL` is also trusted in the
`MEDIA_URL` context, in the same way that the a value trusted as `RESOURCE_URL` is also
trusted in the `URL` context (and transitively also the `MEDIA_URL` context).

The `$sce` service will now automatically attempt to sanitize non-trusted values that
require the `URL` or `MEDIA_URL` context:

* When calling `getTrustedMediaUrl()` a value that is not already a trusted `MEDIA_URL`
will be sanitized using the `imgSrcSanitizationWhitelist`.
* When calling `getTrustedUrl()` a value that is not already a trusted `URL` will be
sanitized using the `aHrefSanitizationWhitelist`.

This results in behaviour that closely matches the previous sanitization behaviour.

To keep rough compatibility with existing apps, we need to allow concatenation of values
that may contain trusted contexts. The following approach is taken for situations that
require a `URL` or `MEDIA_URL` secure context:

* A single trusted value is trusted, e.g. `"{{trustedUrl}}"` and will not be sanitized.
* A single non-trusted value, e.g. `"{{ 'javascript:foo' }}"`, will be handled by
  `getTrustedMediaUrl` or `getTrustedUrl)` and sanitized.
* Any concatenation of values (which may or may not be trusted) results in a
  non-trusted type that will be handled by `getTrustedMediaUrl` or `getTrustedUrl` once the
  concatenation is complete.
  E.g. `"javascript:{{safeType}}"` is a concatenation of a non-trusted and a trusted value,
  which will be sanitized as a whole after unwrapping the `safeType` value.
* An interpolation containing no expressions will still be handled by `getTrustedMediaUrl` or
  `getTrustedUrl`, whereas before this would have been short-circuited in the `$interpolate`
  service. E.g. `"some/hard/coded/url"`. This ensures that `ngHref` and similar directives
  still securely, even if the URL is hard-coded into a template or index.html (perhaps by
  server-side rendering).

BREAKING CHANGES:

If you use `attrs.$set` for URL attributes (a[href] and img[src]) there will no
longer be any automated sanitization of the value. This is in line with other
programmatic operations, such as writing to the innerHTML of an element.

If you are programmatically writing URL values to attributes from untrusted
input then you must sanitize it yourself. You could write your own sanitizer or copy
the private `$$sanitizeUri` service.

Note that values that have been passed through the `$interpolate` service within the
`URL` or `MEDIA_URL` will have already been sanitized, so you would not need to sanitize
these values again.

petebacondarwin commented on src/ng/interpolate.js:
```diff
@@ -238,31 +238,36 @@ function $InterpolateProvider() {
-          expressionPositions = [],
-          singleExpression,
-          contextAllowsConcatenation = trustedContext === $sce.URL || trustedContext === $sce.MEDIA_URL;
+      endIndex,
```
Yes, yes, indentation 😱

petebacondarwin commented on test/ng/directive/booleanAttrsSpec.js:
```diff
@@ -118,221 +118,3 @@ describe('boolean attr directives', function() {
 });
-
-
-describe('ngSrc', function() {
```
All these tests here and below were in the wrong file. I have moved them to their appropriate places.

petebacondarwin commented on test/ng/directive/ngHrefSpec.js:
```diff
@@ -0,0 +1,104 @@
+    element = $compile('<a ng-href="http://server"></a>')($rootScope);
+    $rootScope.$digest();
+    expect(element.attr('href')).toEqual('http://server');
+  }));
```
This is the key new test for this commit

petebacondarwin commented on test/ng/directive/ngHrefSpec.js:
```diff
@@ -0,0 +1,104 @@
+    element = $compile('<a ng-href="javascript:alert(1);">')($rootScope);
+    $rootScope.$digest();
+    expect(element.attr('href')).toBe('unsafe:javascript:alert(1);');
+  }));
```
This is the key new test for this commit

petebacondarwin commented on test/ng/directive/ngSrcSpec.js:
```diff
@@ -18,12 +18,66 @@ describe('ngSrc', function() {
+      element = $compile('<img ng-src="javascript:alert(1);">')($rootScope);
+      $rootScope.$digest();
+      expect(element.attr('src')).toBe('unsafe:javascript:alert(1);');
+    }));
```
And this test too.

gkalpak reviewed (approved):
LGTM (as soon as Travis is green)
